verification-annotations/src/klee.rs

```diff
@@ -60,6 +62,51 @@ make_verifier_nondet!(isize);
 make_verifier_nondet!(f32);
 make_verifier_nondet!(f64);
 
+/// Allocate a symbolic vector of bytes
+pub fn verifier_nondet_bytes(n: usize) -> Vec<u8> {
+    // create an empty vector with capacity 'n'
+    let v: Vec<u8> = Vec::with_capacity(n);
+
+    // take vector apart so that we can access unused capacity
+    let mut v = mem::ManuallyDrop::new(v);
+    let p = v.as_mut_ptr();
+
+    unsafe {
+        // mark contents of vector as symbolic
+        let data = p as *mut raw::c_void;
+        let null = 0 as *const i8;
+        klee_make_symbolic(data, n, null);
+
+        // put vector back together again,
+        // with same capacity but increased length
+        Vec::from_raw_parts(p, n, n)
+    }
+}
+
+/// Allocate a symbolic CString
+pub fn verifier_nondet_cstring(size_excluding_null: usize) -> CString {
+    let mut r = verifier_nondet_bytes(size_excluding_null + 1);
+    for i in 0..size_excluding_null {
+        assume(r[i] != 0u8);
+    }
+    r[size_excluding_null] = 0u8;
+    unsafe { CString::from_vec_with_nul_unchecked(r) }
+}
+
+/// Allocate a symbolic ASCII String
+/// (ASCII strings avoid the complexity of UTF-8)
+pub fn verifier_nondet_ascii_string(n: usize) -> String {
+    let r = verifier_nondet_bytes(n);
+    for i in 0..n {
+        assume(r[i] != 0u8);
+        assume(r[i].is_ascii());
+    }
+    match String::from_utf8(r) {
+        Ok(r) => r,
+        Err(_) => reject(),
```

Comment:
Maybe this doesn't matter very much, but shouldn't this `reject()` be `unreachable()`?

Reply:
reject() is equivalent to assume(false) and so it prevents the verifier from reporting errors involving this path. (In KLEE, it silently halts the current path exploration.)

unreachable() generates an error that is treated as a verification error.

In this case, the function is meant to be generating a legal string so we would not want to report an error.

Reply:
But you already assume all the bytes are ascii, is the Err branch still reachable?

Reply:
Ah, I see.

I am 99.9999% certain that this cannot fail.
But suppose I am wrong, then I would not want to report 'unreachable' because it would be an error in the verification support code not in the code being verified.

It would be good though to get some sort of report that there is a problem in the verification harness.
We should think about this a bit.

I think that this relates to a general problem in verification where you accidentally over-constrain the inputs so that you are not testing inputs that actually matter. In the extreme case (that is surprisingly common), the constraints are contradictory so there are no inputs and the verification is vacuously true.
I have been thinking that we should tackle that using a variation of what hardware engineers do when doing random testing: the System Verilog language lets you gather statistics about how many of the random inputs are in certain ranges - if a given range doesn't have enough values in it, you either need to test more or modify your test.
This would make sense as a proptest extension while for propverify we would just check that each range of interest is satisfiable.
(But exploring this is a problem for the future)